smartsim/types.py

```diff
@@ -0,0 +1,31 @@
+# BSD 2-Clause License
+#
+# Copyright (c) 2021-2024, Hewlett Packard Enterprise
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import annotations
+
+import typing as t
+
+LaunchedJobID = t.NewType("LaunchedJobID", str)
```

Comment:
why did you create a types module?

Reply:
Just in case we have any types that we want to expose to end users. In this case, if a user wrote the following driver script
```py
from smartsim.launchables import Job
from smartsim.entity import Application
from smartsim.settings import LaunchSettings
from smartsim import Experiment

job = Job(
    Application(...),
    LaunchSettings(...)
)

exp = Experiment(...)
ids = exp.start_jobs(job)
reveal_type(ids)  # reveals: tuple[LaunchedJobID, ...]
```
that way users can type hint their driver scripts (on the off chance that they would like to) with out reaching into "protected" modules
```py
from smartsim.types import LaunchedJobID

ids: tuple[LaunchedJobID, ...] | None
if should_start():
    ids = exp.start_jobs(*jobs)
print(f"Launched Jobs: {', '.join(ids)}" if ids is not None else "Nothing Launched :(")
```